### PR TITLE
Publish events tests

### PR DESCRIPTION
**Pull request by qnkhuat**
Adding a bunch of tests around "calling this API should fire an event"

### CODE REVIEW

File: src/metabase/api/card.clj [hunk 1]
```diff
@@ -601,16 +601,16 @@ saved later when it is ready."
 (defn- publish-card-update!
   "Publish an event appropriate for the update(s) done to this CARD (`:card-update`, or archiving/unarchiving
   events)."
-  [card archived?]
+  [card-before-updates card-after-update archived?]
```
Comment (qnkhuat) [inline src/metabase/api/card.clj:604]: this was a bug, even when we archive/unarchive a card, the event "card-updated" were fired.

with this change it should be correct

File: src/metabase/api/card.clj [hunk 2]
```diff
@@ -601,16 +601,16 @@ saved later when it is ready."
...
   (let [event (cond
                 ;; card was archived
                 (and archived?
-                     (not (:archived card))) :card-archive
+                     (not (:archived card-before-updates))) :card-archive
```
Comment (calherries) [inline src/metabase/api/card.clj:608]: I feel like this is more clearly expressed without using the `archived?` arg. You should also be able to get rid of the comments like
```clojure
(and (:archived card-after-updates)
     (not (:archived card-before-updates))) :card-archive
```

File: test/metabase/test.clj [hunk 3]
```diff
@@ -431,14 +431,23 @@
...
+        ;; to avoid flaky tests
+        non-login-events        (fn [] (remove #(#{:user-login} (first %)) @collector))]
     (with-redefs [events/publish-event! (fn [topic event-item]
                                           (swap! collector conj [topic event-item])
                                           (original-publish-event! topic event-item))]
       (assert @#'events/events-initialized?, "events was not initialized")
-      (f {:events     (fn [] @collector)
-          :last-event #(last @collector)}))))
-
+      (f {:raw-events   (fn [] @collector) ;; provide a way to get raw all events
```
Comment (calherries) [inline test/metabase/test.clj:444]: https://guide.clojure.style/#one-semicolon-for-margin-comments
```suggestion
      (f {:raw-events   (fn [] @collector) ; provide a way to get raw all events
```

File: test/metabase/api/dashboard_test.clj [hunk 4]
```diff
@@ -178,14 +178,16 @@
...
 
-    (is (= {:errors {:parameters (str "value may be nil, or if non-nil, value must be an array. "
-                                      "Each parameter must be a map with :id and :type keys")}}
-           (mt/user-http-request :crowberto :post 400 "dashboard" {:name       "Test"
-                                                                   :parameters "abc"})))))
+      (is (= {:errors {:parameters (str "value may be nil, or if non-nil, value must be an array. "
+                                        "Each parameter must be a map with :id and :type keys")}}
+             (mt/user-http-request :crowberto :post 400 "dashboard" {:name       "Test"
+                                                                     :parameters "abc"})))
+      (is (= 0 (count (events)))))))
```
Comment (calherries) [inline test/metabase/api/dashboard_test.clj:190]: Same as
```suggestion
      (is (empty? (events))))))
```
but this has a marginally better error description

File: test/metabase/test.clj [hunk 5]
```diff
@@ -431,14 +431,23 @@
...
+          :events       non-login-events
+          :last-event   #(last (non-login-events))
+          :topic-count  (fn []
```
Comment (calherries) [inline test/metabase/test.clj:447]: `:topic-count` sounds like a single number. I would rename to `topic-counts` or `topic->count`

File: src/metabase/api/card.clj [hunk 6]
```diff
@@ -601,16 +601,16 @@ saved later when it is ready."
 (defn- publish-card-update!
   "Publish an event appropriate for the update(s) done to this CARD (`:card-update`, or archiving/unarchiving
```
Comment (calherries) [inline src/metabase/api/card.clj:602]: :card-archive and :card-unarchive appear unused. Should we just remove them instead?

**Review by calherries — APPROVED** [review]
LGTM, just have a few subjective suggestions

File: src/metabase/api/card.clj (hunk 6 above)
Comment (qnkhuat) [reply src/metabase/api/card.clj:602]: I'd not change that here, I want to keep this PR only for adding tests

File: src/metabase/api/card.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/api/card.clj:608]: it's better! I made the change in 88da822e5d

File: test/metabase/api/dashboard_test.clj (hunk 4 above)
Comment (qnkhuat) [reply test/metabase/api/dashboard_test.clj]: yep 88da822e5d

File: test/metabase/test.clj (hunk 5 above)
Comment (qnkhuat) [reply test/metabase/test.clj]: I went with `topic->count` 88da822e5d

**Comment by qnkhuat** [comment]
we refactored the event this system in https://github.com/metabase/metabase/pull/33392, so this is no longer relevant.

**Pull request closed without merging**